### ISSUE TITLE
Parallel execution of hyperopt on Spark

### ISSUE THREAD

**Issue by asstergi** [hyperopt/hyperopt#414]
Has anyone tried to adapt hyperopt to run in parallel on Spark?

I may need to do so and looking for a starting point.

**Comment by maxpumperla** [comment]
I started doing this with elephas, but it may need some work

**Comment by asstergi** [comment]
@maxpumperla I read on elephas that `right now, the only available optimization algorithm is random search`. Is this still true? Any reasons that TPE was more difficult to implement?

**Comment by maxpumperla** [comment]
@asstergi that is still true. I just wanted to give you a hint on how this could be done. more importantly, elephas only supports keras models. you might want something else.

what you want to do is hijack spark for model and data distribution, compute trials on workers, collect them on master and then send out a new set of trials. The "problem" is that Spark can not easily be used as async job queue. The main paradigm in Spark is to set up an RDD or DataFrame and compute a single operation on distributed data in a synchronous fashion, i.e. your master will get the results when all workers are done. For TPE to work well you want to get back results asap so your new trial selection is based on recent info. To circumvent that, you'd need to set up some sort of parameter server that does this for you (elephas has an experimental HTTP server to do this). MongoDB is better suited for this kind of parallelization in a way, but I agree it would be nice to have a Spark integration for hyperopt.

How exactly do you envision this to go down? How do you want to set up your experiments?

Might be interesting as standalone project for the future. Can be done

**Comment by asstergi** [comment]
@maxpumperla Thank you for all this information! I'm actually now working with Keras models but I definitely need something more generic.

I've quickly checked elephas code and as far as I understand, the code retrieves as many random hyper-parameter configurations as needed and distributes them to the workers to run each model (please correct me if I'm wrong).

I was thinking something along the lines of what you said. To set-up a parameter server and somehow make this work. I haven't thought a lot on this. I'm just checking the waters to see if there's anything that works out of the box.

**Comment by maxpumperla** [comment]
that is correct. keep me posted on what you end up doing, very interesting topic

**Comment by jkbradley** [comment]
@maxpumperla I work with Apache Spark and am interested in contributing this integration feature to hyperopt.  My team has done a bit of prototyping, and I think we can achieve reasonably high parallelism without a heavy-duty addition like a parameter server.

Do you have spare cycles to help review the work?  We can share more info either here or in a design doc.

Also @asstergi were you able to test this out?

Thanks both in advance!

CC @yifan-cao

**Comment by asstergi** [comment]
@jkbradley I haven't tried this out yet. Still, I'd be interested if this would work out.

**Comment by maxpumperla** [comment]
@jkbradley finally caught up with my GH issues, please share what you've got, preferably early so we can ensure we go in the right direction. thanks!

**Comment by jkbradley** [comment]
Here's the design sketch.  We've got some prototype code we can clean up & start sending piecemeal as PRs soon too, pending comments on the design.  The main item we spent time prototyping was different ways of parallelizing via Spark; we decided that concurrent Spark jobs are the simplest method which scales.
https://docs.google.com/document/d/1R-qMVOYxaLw3HdqMiTADQDZ5z7uInMR-x42HFCXu2-s/

Let me know what you think!

**Comment by maxpumperla** [comment]
@jkbradley thanks, I've commented on the doc. It's great that you got around that quickly, I really like your approach. Looking forward to your PRs.

**Comment by jkbradley** [comment]
Thanks for the feedback!  We'll plan time to start this process.  Also, I added a list of planned pieces (PRs) to the end of the doc.

**Comment by ekerazha** [comment]
I'm also interested in this project @jkbradley I can help if needed

**Comment by jkbradley** [comment]
Thanks!  Sorry for the delay.  We hit an unexpected issue with PySpark not handling job cancellation the way its API docs say it does and had to figure out a workaround.  We'll be able to start working on sending PRs soon, with the caveat that SparkTrials will require an updated Spark version (3.0, unless we backport an API into 2.4.x which we probably should not do) for this integration to handle job cancellation properly.

**Comment by maxpumperla** [comment]
@jkbradley given that this is a new integration, I don't see a problem with it requiring a recent Spark version. I assume `pyspark` is kept up to date, so this won't be an issue in practice.

**Comment by pranab** [comment]
How about  splitting the hyperopt config space. Each spark task executes fmin with it's own split of the hyperopt config space. In the driver get the min of all min returned from all tasks/partitions. Although searching fragments of the space and then consolidating may not be as optimum as searching the whole config space, the benefit of parrallelism may offset that.

For example, if I am optimizing over multiple sklearn classification algorithms, each algorithm along with it's config space will be one spark partition/task. In other words, I am splitting the config space along algorithms.

**Comment by jkbradley** [comment]
Now that we've merged SparkTrials, should we mark this issue as complete?

**Comment by maxpumperla** [comment]
good point!
